File: geeta book app/lib/shared/domain/models/activity.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'activity.freezed.dart';
part 'activity.g.dart';

enum ActivityType {
  read,
  watch,
  experiments,
  listen,
  quiz,
  reflection,
  discussion,
  dailyVerse,
}

@freezed
class Activity with _$Activity {
  const factory Activity({
    required String id,
    required String chapterId,
    required String title,
    required String description,
    required ActivityType type,
    required String iconUrl,
    @Default(false) bool isCompleted,
    int? completionPercentage,
    DateTime? lastAccessed,
  }) = _Activity;

  factory Activity.fromJson(Map<String, dynamic> json) =>
      _$ActivityFromJson(json);
}

File: server/src/models/Activity.js
const mongoose = require('mongoose');

const activitySchema = new mongoose.Schema({
  chapterId: {
    type: String,
    required: true,
    index: true,
  },
  title: {
    type: String,
    required: true,
  },
  description: {
    type: String,
    required: true,
  },
  type: {
    type: String,
    enum: ['read', 'watch', 'experiments', 'listen', 'quiz', 'reflection', 'discussion', 'dailyVerse'],
    required: true,
  },
  iconUrl: {
    type: String,
    default: '',
  },
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true },
});

// Virtual for activity ID
activitySchema.virtual('id').get(function() {
  return `${this.chapterId}-${this.type}`;
});

module.exports = mongoose.model('Activity', activitySchema);
